src/modules/notary/index.ts

```diff
@@ -0,0 +1,137 @@
+export type Verifier<T> = (utf8: Buffer, signature: Signature, payload: T) => Promise<boolean>;
+
+export const jsonStringify = <T>(value: T): JsonString<T> => JSON.stringify(value) as JsonString<T>;
+
+export const jsonParse =
+  <T>(validate: Validator<T>): Parser<T> =>
+  (string: JsonString<T>) => {
+    const ret = JSON.parse(string);
+
+    if (validate(ret)) return ret;
+
+    return undefined;
+  };
+
+export interface Notarized<T> {
+  payload: JsonString<T>;
+  signature: Signature;
+}
+
+const defaultFormat = (hash: Buffer) =>
+  `Your transaction fingerprint is: ${hash.toString('base64')}`;
+
+const createMessage = async (payloadStr: string, format: Formatter): Promise<Buffer> => {
+  const payloadBuf = Buffer.from(payloadStr, 'utf-8');
+  const payloadArr = new Uint8Array(
+    payloadBuf.buffer.slice(payloadBuf.byteOffset, payloadBuf.byteOffset + payloadBuf.byteLength)
+  );
+  const bytes = Buffer.from(await sha256(payloadArr, { outputFormat: 'buffer' }));
+
+  return Buffer.from(format(bytes), 'utf-8');
+};
+
+export const signPhantom =
+  (solana: Solana): Signer =>
+  async (utf8) =>
+    (await solana.signMessage(utf8, 'utf-8')).signature as Signature;
+
+export const signNacl =
+  (secret: Uint8Array): Signer =>
+  async (utf8) =>
+    Buffer.from(nacl.sign.detached(utf8, secret)) as Signature;
+
+export const verifyNacl =
+  (pubkey: Uint8Array): Verifier<unknown> =>
+  async (utf8, sig) =>
+    nacl.sign.detached.verify(utf8, sig, pubkey);
+
+export const verifyNaclSelfContained =
+  <T>(pubkey: (val: T) => Uint8Array): Verifier<T> =>
+  async (utf8, sig, payload) =>
+    nacl.sign.detached.verify(utf8, sig, pubkey(payload));
+
+// TODO: nonce
+const nonceBytes = Buffer.from(nacl.randomBytes(4));
+
+export const notarize = async <T>(
+  payload: T,
+  sign: (utf8: Buffer) => Promise<Signature>,
+  options?: {
+    format?: (bytes: Buffer) => string;
+  }
+): Promise<Notarized<T>> => {
+  const payloadStr = jsonStringify(payload);
+  const msg = await createMessage(payloadStr, options?.format ?? defaultFormat);
+  const signature = await sign(msg);
+
+  return { payload: payloadStr, signature };
+};
+
+export const unpackNotarized = async <T>(
+  notarized: Notarized<T> | undefined,
+  verify: (utf8: Buffer, signature: Signature, payload: T) => Promise<boolean>,
+  options: {
+    format?: (bytes: Buffer) => string;
+  } & ({ parse?: undefined; validate: Validator<T> } | { parse: Parser<T> })
+): Promise<T | undefined> => {
+  // For easy composability with parseNotarized
+  if (notarized === undefined) return undefined;
+
+  const { payload, signature } = notarized;
+  const parse = options?.parse ?? jsonParse(options.validate);
+  const payloadDec = parse(payload);
+
+  if (payloadDec === undefined) return undefined;
+
+  const msg = await createMessage(payload, options?.format ?? defaultFormat);
+
+  try {
+    if (!(await verify(msg, signature, payloadDec))) return undefined;
```

Comment:
This is interesting need to create a message in order ot verify?

```suggestion
  const msg = await createMessage(payload, options?.format ?? defaultFormat);

  try {
   cons isVerified = await verify(msg, signature, payloadDec)
    if (!isVerified) return undefined;
```

Reply:
Yeah, it's a weird quirk with the API that Phantom exposes - it produces a signature of the message that can be verified but not decrypted, so the only way I've found to verify signatures from Phantom is to recreate the exact message the user would have seen and cross-check it with their pubkey with `tweetnacl`.  There might be a way to sign an embedded message with Phantom but their docs are kinda terse.